### PR TITLE
Add rule S5148: Authorizing an opened window to access back to the originating window is security-sensitive

### LINKED ISSUE #2621: Add rule S5148: Authorizing an opened window to access back to the originating window is security-sensitive
Implements [RSPEC-5148](https://jira.sonarsource.com/browse/RSPEC-5148) 

For both JS/TS

Report on `window.open` when:
* the first argument is a string literal starting by "http" (external urls)
* and the third argument is missing or doesn't contain `noopener`

```
window.open("https://example.com/dangerous"); // Noncompliant (s5148)
window.open("http://example.com/dangerous"); // Noncompliant (s5148)
window.open("test.html"); // Compliant (s5148)
window.open("https://example.com/dangerous", "windowname", "resizable"); // Noncompliant (s5148)
window.open("https://example.com/dangerous", "windowname", "noopener"); // Compliant (s5148)
window.open("https://example.com/dangerous", "windowname", "resizable,scrollbars,status,noopener"); // Compliant (s5148)
```

### PR DESCRIPTION
**Pull request by ilia-kebets-sonarsource**
Fixes #2621

### CODE REVIEW

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+import { isIdentifier, isMethodCall, isStringLiteral } from '../utils';
+
+const REQUIRED_OPTION = 'noopener';
+const REQUIRED_OPTION_INDEX = 2;
+const URL_INDEX = 0;
+
+export const rule: Rule.RuleModule = {
+  meta: {
+    messages: {
+      issue: 'Make sure not using "noopener" is safe here.',
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:34]: usually we try to describe the message with `messageId`, `issue` is too generic. How about `missingNoopener`?

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    },
+  },
+  create(context: Rule.RuleContext) {
+    return {
+      CallExpression: (node: estree.CallExpression) => {
+        if (!isMethodCall(node)) {
+          return;
+        }
+        const { property, object } = node.callee;
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:43]: I would swap `property` and `object`

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+        if (!(isIdentifier(property, 'open') && isIdentifier(object, 'window'))) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:45]: `!(a && b)` ->  `!a || !b`

Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:45]: I see that you like early return, but IMO here it makes more sense to not have it.

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+          return;
+        }
+        const args = node.arguments;
+        if (args.length >= URL_INDEX && !isUrl(args[URL_INDEX])) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:49]: IMHO you are abusing early return :) it does not improve the readability here.

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    return false;
+  }
+  return argument.value.includes(REQUIRED_OPTION);
+}
+
+function isUrl(argument: estree.Node): boolean {
+  if (!isStringLiteral(argument)) {
+    return false;
+  }
+  try {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:77]: I think it's safe to simply check if string starts with `http`

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+};
+
+function hasRequiredOption(argument: estree.Node) {
+  if (!isStringLiteral(argument)) {
+    return false;
+  }
+  return argument.value.includes(REQUIRED_OPTION);
+}
+
+function isUrl(argument: estree.Node): boolean {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:73]: you can use `getValueOfExpression` from utils to cover more cases. E.g. 
```
let x = "http://hey.com"
window.open(x);
```

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+  if (!isStringLiteral(argument)) {
+    return false;
+  }
+  try {
+    new URL(argument.value);
+    return true;
+  } catch (_) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:80]: you can simply write `} catch {`

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+            node: property,
+          });
+        }
+      },
+    };
+  },
+};
+
+function hasRequiredOption(argument: estree.Node) {
+  if (!isStringLiteral(argument)) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:67]: `getValueOfExpression` can be used here as well

File: eslint-bridge/src/rules/link-with-target-blank.ts (hunk 3 above)
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:45]: this is good for checking `window.open`. I would suggest to support `this.window.open` and then it's good.

You don't need `isWindowObject` here as it's checking if variable has [`Window`](https://developer.mozilla.org/en-US/docs/Web/API/Window) type

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/rules/link-with-target-blank.ts (hunk 5 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/src/rules/link-with-target-blank.ts:77]: OK, so the current implementation is wrong as it flags `file://...`
However, there might be a file starting with `http`, like `httpprocol.png`.
Therefore I propose to check if it starts with `http://` or `https://`.
@vilchik-elena WDYT?

Comment (vilchik-elena) [reply eslint-bridge/src/rules/link-with-target-blank.ts:77]: sounds good

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+            messageId: 'missingNoopener',
+            node: property,
+          });
+        }
+      },
+    };
+  },
+};
+
+function isThisWindow(node: estree.Node) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:69]: there is `isMemberExpression` in utils

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  create(context: Rule.RuleContext) {
+    return {
+      CallExpression: (node: estree.CallExpression) => {
+        if (!isMethodCall(node)) {
+          return;
+        }
+        const { object, property } = node.callee;
+
+        if (
+          !isIdentifier(property, 'open') ||
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:45]: if you use early return then make the condition as simple as possible. Here you can extract it into another function or variable (to have opportunity to name the condition)

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+          (!isIdentifier(object, 'window') && !isThisWindow(object))
+        ) {
+          return;
+        }
+
+        const args = node.arguments;
+        if (args.length >= URL_INDEX && !isHttpUrl(context, args[URL_INDEX])) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:52]: same here:
```suggestion
        const hasHttpUrl = URL_INDEX < args.length && isHttpUrl(context, args[URL_INDEX]);
        if (!hasHttpUrl) {
```

Note that I didn't invert `&&`. That way we will also return early if you have 0 arguments.

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  return (
+    node.type === 'MemberExpression' &&
+    node.object.type === 'ThisExpression' &&
+    isIdentifier(node.property, 'window')
+  );
+}
+
+function hasRequiredOption(context: Rule.RuleContext, argument: estree.Node) {
+  const literalNode = getValueOfExpression(context, argument, 'Literal');
+  if (literalNode === undefined || !isStringLiteral(literalNode)) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:79]: `literalNode === undefined` -> `!literalNode`

File: eslint-bridge/src/rules/link-with-target-blank.ts [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    return false;
+  }
+  return literalNode.value.includes(REQUIRED_OPTION);
+}
+
+function isHttpUrl(context: Rule.RuleContext, argument: estree.Node): boolean {
+  const literalNode = getValueOfExpression(context, argument, 'Literal');
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/link-with-target-blank.ts:86]: how about removing duplication here and above?

File: eslint-bridge/src/rules/post-message.ts [hunk 14]
```diff
@@ -57,12 +57,6 @@ export const rule: Rule.RuleModule = {
   },
 };
 
-function isWindowObject(node: estree.Node, context: Rule.RuleContext) {
```
Comment (vilchik-elena) [inline eslint-bridge/src/rules/post-message.ts:60]: bring all this back

File: eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts [hunk 15]
```diff
@@ -0,0 +1,36 @@
+window.open('https://example.com/dangerous1'); // Noncompliant {{Make sure not using "noopener" is safe here.}}
+    // ^^^^
+window.open('http://example.com/dangerous2'); // Noncompliant
+    // ^^^^
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts:4]: IMO no need to assert location for every issue

File: eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+window.open('https://example.com/dangerous3', 'windowname', 'resizable'); // Noncompliant
+    // ^^^^
+window.open('https://example.com/dangerous4', 'windowname', 123); // Noncompliant
+    // ^^^^
+
+window.open('test.html');
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts:10]: add test with 0 arguments

File: eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts [hunk 17]
```diff
@@ -0,0 +1,36 @@
...
+const requiredOption = 'noopener';
+const missingRequiredOption = 'resizable';
+window.open(httpUrl, 'windowname', missingRequiredOption); // Noncompliant
+    // ^^^^
+window.open(otherUrl, 'windowname', missingRequiredOption);
+
+window.open(httpUrl, 'windowname', requiredOption);
+
+this.window.open('https://example.com/dangerous10', 'windowname', 'resizable'); // Noncompliant
+         // ^^^^
```
Comment (vilchik-elena) [inline eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts:36]: new line ;)
Just review your changes in github before you assign for review

File: javascript-checks/src/main/java/org/sonar/javascript/checks/LinkWithTargetBlankCheck.java [hunk 18]
```diff
@@ -0,0 +1,38 @@
...
+import org.sonar.plugins.javascript.api.EslintBasedCheck;
+import org.sonar.plugins.javascript.api.JavaScriptRule;
+import org.sonar.plugins.javascript.api.TypeScriptRule;
+
+@JavaScriptRule
+@TypeScriptRule
+@Rule(key = "S5148")
+public class LinkWithTargetBlankCheck implements EslintBasedCheck{
+
+  public static final boolean CONSIDER_JSDOC = false;
```
Comment (vilchik-elena) [inline javascript-checks/src/main/java/org/sonar/javascript/checks/LinkWithTargetBlankCheck.java:32]: huh?

**Review by vilchik-elena — CHANGES_REQUESTED** [review]

File: eslint-bridge/src/rules/link-with-target-blank.ts (hunk 10 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/src/rules/link-with-target-blank.ts:45]: `extract it into another function or variable (to have the opportunity to name the condition)` awesome!

File: eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts (hunk 17 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts:36]: I need to setup my IDE to flag this.

File: eslint-bridge/src/rules/link-with-target-blank.ts (hunk 9 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/src/rules/link-with-target-blank.ts:72]: I'm not sure how you propose to refactor it.

File: eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+window.open('https://example.com/dangerous6', 'windowname', 'resizable,scrollbars,status,noopener');
+
+function open() {
+    console.log('hello')
+}
+open('https://example.com/dangerous7', 'windowname', 'resizable');
+
+window.notOpen('https://example.com/dangerous8', 'windowname', 'resizable');
+
+const httpUrl = 'https://example.com/dangerous9';
```
[the remark this answers is not in the document]
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts:24]: I added https://github.com/SonarSource/SonarJS/pull/3167/files#diff-d110b60ab02d53976ba21ce2ff17b333ed2e97beae59f6b886cc11b69244634aR36

Not sure it's what you suggested.

File: javascript-checks/src/main/java/org/sonar/javascript/checks/LinkWithTargetBlankCheck.java (hunk 18 above)
Comment (ilia-kebets-sonarsource) [reply javascript-checks/src/main/java/org/sonar/javascript/checks/LinkWithTargetBlankCheck.java]: mistake when copying

File: eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts (hunk 17 above)
Comment (ilia-kebets-sonarsource) [reply eslint-bridge/tests/rules/comment-based/link-with-target-blank.ts]: changed it in VScode

File: eslint-bridge/src/rules/link-with-target-blank.ts (hunk 9 above)
Comment (vilchik-elena) [reply eslint-bridge/src/rules/link-with-target-blank.ts:72]: indeed you are right

**Review by vilchik-elena — APPROVED** [review]
Well done!

**Pull request merged**